packages/compiler/emitter-framework/type-emitter.ts

```diff
@@ -0,0 +1,745 @@
+import {
+  BooleanLiteral,
+  compilerAssert,
+  Enum,
+  EnumMember,
+  Interface,
+  IntrinsicType,
+  isTemplateDeclaration,
+  Model,
+  ModelProperty,
+  Namespace,
+  NumericLiteral,
+  Operation,
+  Program,
+  Scalar,
+  StringLiteral,
+  Tuple,
+  Type,
+  Union,
+  UnionVariant,
+} from "../core/index.js";
+import { code, StringBuilder } from "./builders/string.js";
+import { Placeholder } from "./placeholder.js";
+import {
+  AssetEmitter,
+  CadlDeclaration,
+  Context,
+  Declaration,
+  EmitEntity,
+  EmittedSourceFile,
+  Scope,
+  SourceFile,
+} from "./types.js";
+
+export type EmitterOutput<T> = EmitEntity<T> | Placeholder<T> | T;
+
+/**
+ * Implement emitter logic by extending this class and passing it to
+ * `emitContext.createAssetEmitter`. This class should not be constructed
+ * directly.
```

Comment:
Should it be marked `abstract` then?

Reply:
It turns out the answer to this is no, because it causes significant type complexity having to do with the fact that functions which take a `typeof TypeEmitter` as a parameter (like createAssetEmitter)... can't anymore, because the compiler complains that it's abstract. The general fix (I THINK) is to replace such instances with `new<T>(assetEmitter: AssetEmitter<T>): TypeEmitter<T>`, but that was having deleterious effects all over the codebase. I'm giving up.